### PR TITLE
Make non_linear interpolation independent of a chosen value type

### PR DESCRIPTION
**Pull request by wdeconinck**
This PR addresses issue #173 

@twsearle please check if this suits your needs.

### CODE REVIEW

File: src/atlas/interpolation/nonlinear/Missing.h [hunk 1]
```diff
@@ -29,9 +30,19 @@ struct Missing : NonLinear {
 };
 
 
-template <typename T>
 struct MissingIfAllMissing : Missing {
     bool execute(NonLinear::Matrix& W, const Field& field) const {
```
Comment (pmaciel) [inline src/atlas/interpolation/nonlinear/Missing.h:34]: Only missing to move the ::execute methods (MissingIfAllMissing::, MissingIfAnyMissing::, MissingIfHeaviestMissing::execute) to the .cc

**Review by pmaciel — APPROVED** [review]
Looks great! I saw the issue coming in, and I'm glad it's already solved :-)

File: src/atlas/interpolation/nonlinear/Missing.h (hunk 1 above)
Comment (wdeconinck) [reply src/atlas/interpolation/nonlinear/Missing.h:34]: Good point! Will do.

File: src/tests/interpolation/test_interpolation_non_linear.cc [hunk 2]
```diff
@@ -205,6 +205,81 @@ CASE("Interpolation of rank 2 field with MissingValue") {
...
+
+    auto interpolate = [&](const std::string& missing_type) {
+        Config config;
+        config.set("type", "finite-element");
+        config.set("non_linear", missing_type);
+        Interpolation interpolation(config, fsA, fsB);
+        interpolation.execute(fieldsA, fieldsB);
+    };
+
+    SECTION( "missing-if-any-missing" ) {
```
[the remark this answers is not in the document]
Comment (pmaciel) [reply src/tests/interpolation/test_interpolation_non_linear.cc:269]: Oh right -- I read it too quickly! It looks great actually

File: src/atlas/interpolation/nonlinear/Missing.h (hunk 1 above)
Comment (wdeconinck) [reply src/atlas/interpolation/nonlinear/Missing.h]: Done now with [b32c316](https://github.com/ecmwf/atlas/pull/176/commits/b32c316e3f31c011ab67688a09bcea7fb3ef016d)

**Comment by twsearle** [comment]
Lovely - this looks exactly like I was imagining! Unfortunately I am having environment issues this week so I haven't been able to test it on my problems but I am confident.

**Comment by twsearle** [comment]
I have tested this now and it solves my problem thanks

**Comment by twsearle** [comment]
Could we possibly get this one merged in before the next release @wdeconinck ?

**Comment by wdeconinck** [comment]
Hi @twsearle sorry it was an oversight that it was not merged in previous version. It is merged now in develop and will be in the next version.

**Pull request merged**